### PR TITLE
Add limit_all_gathers option to fsdp_config and fix forward_prefetch bug

### LINKED ISSUE #21156: [HF Trainer] [PyTorch FSDP] Add support for backward_prefetch, forward_prefetch, limit_all_gathers
### **Feature request**
Can we add Trainer support for the following [FSDP](https://pytorch.org/docs/1.13/fsdp.html?highlight=fsdp#module-torch.distributed.fsdp) features? `backward_prefetch`, `forward_prefetch` and  `limit_all_gathers`

### **Motivation**

`backward_prefetch`, `forward_prefetch` and  `limit_all_gathers` are important to achieve best performance when training at scale with FSDP.

### PR DESCRIPTION
**Pull request by raghavanone**
fixes #21156 
Add limit_all_gathers option to fsdp_config.
Fix forward_prefetch bug.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -322,8 +322,9 @@ def __init__(
             logger.info(f"No `TrainingArguments` passed, using `output_dir={output_dir}`.")
             args = TrainingArguments(output_dir=output_dir)
         self.args = args
-        # Seed must be set before instantiating the model when using model
-        enable_full_determinism(self.args.seed) if self.args.full_determinism else set_seed(self.args.seed)
+        (  # Seed must be set before instantiating the model when using model
+            enable_full_determinism(self.args.seed) if self.args.full_determinism else set_seed(self.args.seed)
+        )
```
Comment (sgugger) [inline src/transformers/trainer.py:327]: This change makes no sense and should be reverted.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -348,9 +349,11 @@ def __init__(
...
             if model_init is not None:
                 warnings.warn(
-                    "`Trainer` requires either a `model` or `model_init` argument, but not both. `model_init` will"
-                    " overwrite your model when calling the `train` method. This will become a fatal error in the next"
-                    " release.",
+                    (
+                        "`Trainer` requires either a `model` or `model_init` argument, but not both. `model_init` will"
+                        " overwrite your model when calling the `train` method. This will become a fatal error in the"
+                        " next release."
+                    ),
```
Comment (sgugger) [inline src/transformers/trainer.py:356]: Same here.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -441,9 +444,13 @@ def __init__(
                 self.backward_prefetch = BackwardPrefetch.BACKWARD_POST
 
             self.forword_prefetch = False
-            if "forword_prefetch" in self.args.fsdp_config and self.backward_prefetch:
+            if "forword_prefetch" in self.args.fsdp_config and self.forword_prefetch:
                 self.forword_prefetch = True
```
Comment (sgugger) [inline src/transformers/trainer.py:448]: This test can never be satisfied. Maybe you meant `self.args.fsdp_config.get("forword_prefect", False)`? (would also be nice to fix the typo)

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -441,9 +444,13 @@ def __init__(
...
 
+            self.limit_all_gathers = False
+            if "limit_all_gathers" in self.args.fsdp_config and self.limit_all_gathers:
+                self.limit_all_gathers = True
```
Comment (sgugger) [inline src/transformers/trainer.py:452]: Same here this test can never be satisfied.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -1530,8 +1538,10 @@ def train(
         if "model_path" in kwargs:
             resume_from_checkpoint = kwargs.pop("model_path")
             warnings.warn(
-                "`model_path` is deprecated and will be removed in a future version. Use `resume_from_checkpoint` "
-                "instead.",
+                (
+                    "`model_path` is deprecated and will be removed in a future version. Use `resume_from_checkpoint` "
+                    "instead."
+                ),
```
Comment (sgugger) [inline src/transformers/trainer.py:1544]: Shouldn't be changed.

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -1543,8 +1553,9 @@ def train(
         # Model re-init
         model_reloaded = False
         if self.model_init is not None:
-            # Seed must be set before instantiating the model when using model_init.
-            enable_full_determinism(self.args.seed) if self.args.full_determinism else set_seed(self.args.seed)
+            (  # Seed must be set before instantiating the model when using model_init.
+                enable_full_determinism(self.args.seed) if self.args.full_determinism else set_seed(self.args.seed)
+            )
```
Comment (sgugger) [inline src/transformers/trainer.py:1558]: Same here.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR. There are several changes that are unwarranted in the formatting, make sure to rebase your PR on the main branch and install the right versions of our quality tools with `pip install -e .["quality"]` in the transformers repo.

There are also weird things in the tests you change/introduce that are always false.

**Comment by raghavanone** [comment]
@sgugger  All changes resolved.

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -441,9 +441,13 @@ def __init__(
                 self.backward_prefetch = BackwardPrefetch.BACKWARD_POST
 
             self.forword_prefetch = False
-            if "forword_prefetch" in self.args.fsdp_config and self.backward_prefetch:
+            if self.args.fsdp_config.get("forword_prefect", False):
```
Comment (sgugger) [inline src/transformers/trainer.py:444]: Just checking here with @pacman100 to make user using `backward_prefetch` was a mistake

**Review by sgugger — APPROVED** [review]
Thanks for iterating!

File: src/transformers/trainer.py (hunk 7 above)
Comment (pacman100) [reply src/transformers/trainer.py:444]: yes, this fix is correct

**Review by pacman100 — APPROVED** [review]
Thanks @raghavanone  for enhancement and fixes!

**Pull request merged**